Write about our continuous release process

alexwlchan opened the pull request:
I promised @DRMacIver a post about this at PyCon UK… in 2017. That’s nearly 60 releases ago! And then I was CamPUG this evening, and Tibbs was asking for the post about this, so here we go.

Current status: first draft written at the end of a long day. Definitely needs a good edit, and I’d be interested in hearing whether there are benefits I’ve forgotten that I should write about.

alexwlchan committed:
Initial post on continuous releases

alexwlchan committed:
Bump the date

alexwlchan committed:
Markups and edits

alexwlchan commented:
Okay, I think this is ready for somebody other than me to review.

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+---
+layout: post
+tags: development-process
```
Due to an annoying limitation of Jekyll on Github Pages, our tags are not entirely freeform and new tags need to be added in https://github.com/HypothesisWorks/HypothesisWorks.github.io/tree/master/articles

(there should probably be a script for this)

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+We've completely automated the process of releasing, so every pull request that changes code gets a new release, without any human input.
+In this post, I'll explain how our continuous releases work, and why we find it so useful.
+
+[changelog]: https://hypothesis.readthedocs.io/en/latest/changes.html
```
Huh. I didn't know about this markdown link syntax.

alexwlchan replied:
Yeah, it's quite nice for including links without cluttering up your prose with URLs.

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+In the past, Hypothesis was released manually.
+Somebody had to write a changelog, tag a new release on GitHub, and run some manual pip commands to publish a new version to PyPI -- and only David had the credentials for the latter.
+
+This meant that releases were somewhat infrequent, and features spent a long time in master before they were available to `pip install`.
```
Minor historical note: In earlier history, Hypothesis releases were actually *really* fast - possibly faster than they are today due to much more limited testing. The problem seemed to be that as soon as the release cycle stopped being virtually instantaneous it ground to a halt. I'm not sure exactly how and why this happened, but I think it was partly due to improved testing and process, so the previous YOLO deployments stopped being viable.

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+Somebody had to write a changelog, tag a new release on GitHub, and run some manual pip commands to publish a new version to PyPI -- and only David had the credentials for the latter.
+
+This meant that releases were somewhat infrequent, and features spent a long time in master before they were available to `pip install`.
+The pace of development picked up in 2017 -- partly as new maintainers arrived, and partly work for David's PhD -- and we wanted to be able to release more frequently.
```
Maybe clarify that the PhD hadn't actually started then? e.g. "groundwork for David's upcoming (now started) PhD"? (Not at all wed to that wording)

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+The pace of development picked up in 2017 -- partly as new maintainers arrived, and partly work for David's PhD -- and we wanted to be able to release more frequently.
+We decided to automate the entire release process.
+
+Now, when you create a pull request that changes the Hypothesis code -- anything that gets installed by pip, not the tests or docs -- you have to include a `RELEASE.rst` file, which describes your change.
```
Possibly worth noting that this is the second gen automation system - the first required updating the changelog yourself and was awful.

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+
+I've started using this in my other repos -- both these scripts exactly, and derivatives of the same idea.
+
+If you'd like to try this yourself (and I'd really encourage you to do so!), all the scripts for this process are under the same MPL license as Hypothesis itself.
```
It might be worth noting that the code is a little hairy and will probably require some adaptation...

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+Moving to continuous releases has been amazing.
+
+Our releases happen much more quickly.
+Rather than waiting for somebody to do it manually, every patch is available as soon as our tests confirm it's okay.
```
Another unmentioned advantage is that I *really* hated doing these manual releases! Collecting the changelog manually was really annoying, and we didn't really have a previously good system of change log automation.

DRMacIver commented on _posts/2018-02-12-continuous-releases.md:
```diff
@@ -0,0 +1,107 @@
+The rest is a description of the changes in your patch.
+
+We have a test in CI that checks for this file -- any change to the core code needs a release file, even [fixing a typo][typo].
+If you need a release file but haven't written one, the tests fail and your pull request won't be merged.
```
Maybe worth mentioning that you can create a `RELEASE.rst` without source changes and we sometimes do for e.g. making `setup.py` changes (which maybe we should just add to the checks TBH) and test changes for the benefit of downstream packagers?

DRMacIver reviewed (changes requested):
Generally looks 💯. Have added some comments on things I think could use clarifying or elaborating on.

Would it be worth talking about some of the limitations and edge cases?

e.g.

* This blocks merges of new features until the old one has been released, by virtue of the conflicting adds.
* If a non-feature PR has been merged during the build, the push of the new commit will fail and the deployment will fail, but the build of the new merged version of master will deploy it.

alexwlchan committed:
Create pages for the new tag

alexwlchan committed:
Tidy up the section on automation benefits

alexwlchan committed:
More review markups

alexwlchan committed:
Add a note about scale

alexwlchan commented:
@DRMacIver I think I’ve addressed all your comments, but I want to go over it with a red pen and paper before the next review. I'll let you know when it's ready for a second review.

dchudz commented:
I wonder if it might be worth saying something about how this workflow interacts with supporting old versions. (Hypothesis doesn't ever add a release to old major or minor versions, right?)

I think a project wouldn't have much trouble amending this workflow to support e.g. one old major version. But doing much more than that seems like it would be bad news. This all seems to work really well for Hypothesis but I do wonder about projects with e.g. critical security patches. 

It's a great post even without saying anything about this - I'm not really saying you *should* address it. I guess I'm just mentioning the question because it's one I think readers will ask.

DRMacIver commented:
> I wonder if it might be worth saying something about how this workflow interacts with supporting old versions. (Hypothesis doesn't ever add a release to old major or minor versions, right?)

This isn't a thing we do, so it's not ever really came up.

If anything though I imagine it would make it easier (modulo some wrangling of git): If you apply a patch with a `RELEASE.rst` to an older version, the version number automation would automatically do the right thing.

Hard to say for sure though. I'd be reluctant to suggest much about this without having actually tried such a workflow.

dchudz commented:
> I'd be reluctant to suggest much about this without having actually tried such a workflow.

Yeah, that makes sense. Sorry, I think I really should have just suggested saying something like that: "We like this workflow, but haven't tried it for ... so your mileage may vary." But I guess it kinda goes without saying too. Maybe it wasn't worth mentioning, sorry!

alexwlchan commented:
> I wonder if it might be worth saying something about how this workflow interacts with supporting old versions.

I think the answer is “it doesn’t, at least not in its current form”. 😜 

I’ll put a sentence about it in the “this won’t necessarily work for larger projects” paragraph – I think this is a good point, and we don't need to offer solutions to say that it's a problem.

alexwlchan committed:
More markups on the post

alexwlchan commented:
Okay, markups made. I think this is ready for another look @DRMacIver.

DRMacIver commented:
Looks great!

I'll leave it up to you to merge whenever you want. Sorry for the delayed rereview.

alexwlchan commented:
I’ll need to fix the date (we should totes do continuous deployment here as well!) when I get home; I’ll probably post on Monday morning.

alexwlchan commented:
@DRMacIver Also it looks like I don't have write access on this repo.

alexwlchan committed:
continuous-releases: bump the date on the post

alexwlchan commented:
@DRMacIver I’ve bumped the post date. At this point either you can give me write access, or merge it immediately.

DRMacIver commented:
> @DRMacIver I’ve bumped the post date. At this point either you can give me write access, or merge it immediately.

I went with "both"
